subsys/portability/cmsis_rtos_v2/thread.c

```diff
@@ -477,6 +488,37 @@ osStatus_t osThreadJoin(osThreadId_t thread_id)
 	return (ret == 0) ? osOK : osErrorResource;
 }
 
+/**
+ * @brief Decrement the thread count based on thread type.
+ */
+int decrement_thread_counter(struct cmsis_rtos_thread_cb *tid)
+{
+	if (tid == NULL) {
+		return -EINVAL;
+	}
+
+#if CONFIG_CMSIS_V2_THREAD_DYNAMIC_MAX_COUNT != 0
+	for (int stack_idx = 0; stack_idx < CONFIG_CMSIS_V2_THREAD_DYNAMIC_MAX_COUNT; stack_idx++) {
+		if (tid->z_thread.stack_info.start ==
+		    (uintptr_t)cmsis_rtos_thread_stack_pool[stack_idx]) {
+			atomic_dec(&num_dynamic_stack);
+			stack_status[stack_idx].is_in_use = false;
+			break;
+		}
+	}
+#endif
+
+	for (int cb_idx = 0; cb_idx < CONFIG_CMSIS_V2_THREAD_MAX_COUNT; cb_idx++) {
+		if (tid == &cmsis_rtos_thread_cb_pool[cb_idx]) {
+			atomic_dec(&num_dynamic_cb);
+			break;
+		}
+	}
+
+	atomic_dec((atomic_t *)&thread_num);
+	return 0;
```

Comment:
These changes about `thread_num` and `num_dynamic_cb`. Don't seems directly related to dynamic threads. I believe it makes sense to place that in a separated commit. Then, it will be easier to explain why you need to change it.

(style note: when the loop are simple, call the loop indexes `i` and `j`. Every developers understand `i` and `j` are loop indexes).